File: clients/client-py/queen/http/http_client.py
"""
HTTP client with retry, load balancing, and failover support
"""

import asyncio
import random
from typing import Any, Dict, Optional, Tuple, Union
import httpx

from ..utils import logger
from .load_balancer import LoadBalancer


class HttpClient:
    """HTTP client with retry, load balancing, and failover support"""

    def __init__(
        self,
        *,
        base_url: Optional[str] = None,
        load_balancer: Optional[LoadBalancer] = None,
        timeout_millis: int = 30000,
        retry_attempts: int = 3,
        retry_delay_millis: int = 1000,
        enable_failover: bool = True,
        bearer_token: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        retry_429: Optional[Dict[str, Any]] = None,
        transport: Optional[httpx.BaseTransport] = None,
    ):
        """
        Initialize HTTP client

        Args:
            base_url: Base URL for single server
            load_balancer: LoadBalancer instance for multiple servers
            timeout_millis: Request timeout in milliseconds
            retry_attempts: Number of retry attempts
            retry_delay_millis: Initial retry delay (exponential backoff)
            enable_failover: Enable automatic failover
            bearer_token: Bearer token for proxy authentication
            headers: Custom headers to include in every request
            retry_429: Backoff policy for HTTP 429 (rate limited) responses,
                separate from retry_attempts/retry_delay_millis above. See
                queen.types.Retry429Config for the shape and defaults.
                max_attempts defaults to 10 for ordinary requests; a
                long-poll pop (retry_kind="pop") retries unboundedly unless
                max_attempts is set here, in which case it applies to both.
                PLAN_QUEEN_PROXY_CLOUD.md §4/§9 (client 429 backoff, B4).
            transport: Optional httpx transport override (e.g.
                httpx.MockTransport) for tests -- no broker/network required.
                None uses httpx's normal transport.
        """
        self._base_url = base_url
        self._load_balancer = load_balancer
        self._timeout_millis = timeout_millis
        self._retry_attempts = retry_attempts
        self._retry_delay_millis = retry_delay_millis
        self._enable_failover = enable_failover
        self._bearer_token = bearer_token
        self._retry_429: Dict[str, Any] = dict(retry_429) if retry_429 else {}

        # Build headers with optional auth
        merged_headers: Dict[str, str] = {}
        if bearer_token:
            merged_headers["Authorization"] = f"Bearer {bearer_token}"
        if headers:
            merged_headers.update(headers)

        # Create httpx.AsyncClient (persistent connection pool)
        self._client = httpx.AsyncClient(
            timeout=httpx.Timeout(timeout_millis / 1000.0),
            limits=httpx.Limits(max_keepalive_connections=10, max_connections=100),
            headers=merged_headers,
            transport=transport,
        )

        logger.log(
            "HttpClient.constructor",
            {
                "has_load_balancer": load_balancer is not None,
                "base_url": base_url or "load-balanced",
                "timeout_millis": timeout_millis,
                "retry_attempts": retry_attempts,
                "enable_failover": enable_failover,
                "has_auth": bearer_token is not None,
                "custom_headers": len(headers) if headers else 0,
                "retry_429": self._retry_429,
            },
        )

    def _retry429_policy_for(self, retry_kind: Optional[str]) -> Tuple[Optional[int], int, int]:
        """Resolve the effective 429 retry policy for a request kind.

        - "pop": long-poll pop (wait=True). Unbounded attempts (None) by
          default -- the long-poll loop is meant to keep waiting through
          transient rate limiting -- unless retry_429["max_attempts"] was
          explicitly configured.
        - anything else (push, admin calls, non-waiting pop, ...): bounded,
          defaults to 10 attempts.

        base_ms/cap_ms always default to 500/30000 and apply to both kinds.
        Returns (max_attempts_or_None, base_ms, cap_ms).
        """
        cfg = self._retry_429 or {}
        base_ms = cfg.get("base_ms", 500)
        cap_ms = cfg.get("cap_ms", 30000)
        configured_max = cfg.get("max_attempts")
        if configured_max is not None:
            max_attempts = configured_max
        else:
            max_attempts = None if retry_kind == "pop" else 10
        return max_attempts, base_ms, cap_ms

    @staticmethod
    def _compute_retry429_delay(
        attempt_index: int,
        retry_after_seconds: Optional[float],
        base_ms: int,
        cap_ms: int,
    ) -> float:
        """Delay (seconds, for asyncio.sleep) before the next 429 retry.

        Honors Retry-After (seconds) when the server sent one, with +-20%
        jitter to avoid a synchronized thundering herd; otherwise falls back
        to exponential backoff (base_ms * 2^attempt, capped at cap_ms), also
        jittered +-20%.
        """
        if retry_after_seconds is not None and retry_after_seconds >= 0:
            delay_ms = retry_after_seconds * 1000.0
        else:
            delay_ms = min(cap_ms, base_ms * (2 ** attempt_index))
        jitter_multiplier = 1 + random.uniform(-0.2, 0.2)
        return max(0.0, (delay_ms * jitter_multiplier) / 1000.0)

    async def _execute_with_retry429(
        self,
        url: str,
        method: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
        retry_kind: Optional[str],
    ) -> Any:
        """Run a single logical request against one URL, transparently
        retrying HTTP 429 responses with backoff until the policy for
        `retry_kind` is exhausted (or never, for an unbounded pop policy).
        Any other exception is raised straight through -- 429 is the only
        status this layer treats as retryable; 5xx/network retry and
        cross-backend failover are the caller's job.
        """
        max_attempts, base_ms, cap_ms = self._retry429_policy_for(retry_kind)
        tries = 0
        while True:
            tries += 1
            try:
                return await self._execute_request(url, method, body, request_timeout_millis)
            except httpx.HTTPStatusError as error:
                if error.response.status_code != 429:
                    raise

                if max_attempts is not None and tries >= max_attempts:
                    logger.error(
                        "HttpClient.retry429",
                        {
                            "method": method,
                            "url": url,
                            "error": "max 429 attempts exhausted",
                            "attempts": tries,
                            "code": getattr(error, "code", None),
                        },
                    )
                    raise

                delay = self._compute_retry429_delay(
                    tries - 1, getattr(error, "retry_after_seconds", None), base_ms, cap_ms
                )
                logger.warn(
                    "HttpClient.retry429",
                    {
                        "method": method,
                        "url": url,
                        "attempt": tries,
                        "retry_kind": retry_kind or "default",
                        "next_delay_s": delay,
                        "retry_after_seconds": getattr(error, "retry_after_seconds", None),
                        "code": getattr(error, "code", None),
                    },
                )
                await asyncio.sleep(delay)

    async def get(
        self,
        path: str,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
        *,
        retry_kind: Optional[str] = None,
    ) -> Any:
        """GET request.

        retry_kind: pass "pop" for long-poll (wait=True) pop requests to get
        the unbounded-with-backoff 429 policy; omit for everything else
        (push, admin calls, non-waiting pop), which get the bounded default
        (10 attempts).
        """
        return await self._request_with_failover(
            "GET", path, None, request_timeout_millis, affinity_key, retry_kind
        )

    async def post(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
        *,
        retry_kind: Optional[str] = None,
    ) -> Any:
        """POST request"""
        return await self._request_with_failover(
            "POST", path, body, request_timeout_millis, affinity_key, retry_kind
        )

    async def put(
        self,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
        *,
        retry_kind: Optional[str] = None,
    ) -> Any:
        """PUT request"""
        return await self._request_with_failover(
            "PUT", path, body, request_timeout_millis, affinity_key, retry_kind
        )

    async def delete(
        self,
        path: str,
        request_timeout_millis: Optional[int] = None,
        affinity_key: Optional[str] = None,
        *,
        retry_kind: Optional[str] = None,
    ) -> Any:
        """DELETE request"""
        return await self._request_with_failover(
            "DELETE", path, None, request_timeout_millis, affinity_key, retry_kind
        )

    async def _execute_request(
        self,
        url: str,
        method: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
    ) -> Any:
        """Execute single HTTP request"""
        effective_timeout = request_timeout_millis or self._timeout_millis
        logger.log(
            "HttpClient.request",
            {"method": method, "url": url, "has_body": body is not None, "timeout": effective_timeout},
        )

        try:
            # Set timeout for this specific request
            timeout = httpx.Timeout(effective_timeout / 1000.0)

            # Prepare request
            kwargs: Dict[str, Any] = {
                "method": method,
                "url": url,
                "timeout": timeout,
            }

            if body:
                kwargs["json"] = body

            response = await self._client.request(**kwargs)

            logger.log("HttpClient.response", {"method": method, "url": url, "status": response.status_code})

            # Handle 204 No Content
            if response.status_code == 204:
                return None

            # Handle errors
            if not response.is_success:
                error_msg = f"HTTP {response.status_code}: {response.reason_phrase}"
                code: Optional[str] = None
                try:
                    text = response.text
                    if text:
                        body_data = response.json()
                        error_msg = body_data.get("error", error_msg)
                        code = body_data.get("code")
                except Exception:
                    pass

                logger.error(
                    "HttpClient.request",
                    {"method": method, "url": url, "status": response.status_code, "error": error_msg, "code": code},
                )
                error = httpx.HTTPStatusError(error_msg, request=response.request, response=response)
                # Proxy error contract: 429 {"error", "code": "rate_limited" |
                # "quota_exceeded"} with Retry-After (seconds); 403 {"error",
                # "code": "cluster_suspended" | "storage_quota_exceeded" |
                # "feature_gated" | "forbidden"}. These attributes are always
                # set (code may be None) so callers can branch without
                # string-matching the message.
                error.code = code  # type: ignore[attr-defined]
                error.retry_after_seconds = None  # type: ignore[attr-defined]
                if response.status_code == 429:
                    retry_after = response.headers.get("retry-after")
                    if retry_after is not None:
                        try:
                            error.retry_after_seconds = float(retry_after)  # type: ignore[attr-defined]
                        except (TypeError, ValueError):
                            error.retry_after_seconds = None  # type: ignore[attr-defined]
                raise error

            # Parse successful response
            content_type = response.headers.get("content-type", "")
            content_length = response.headers.get("content-length", "")

            if (
                not content_type
                or "application/json" not in content_type
                or content_length == "0"
            ):
                text = response.text
                if not text:
                    return None
                try:
                    return response.json()
                except Exception:
                    return None

            return response.json()

        except httpx.TimeoutException as e:
            logger.error(
                "HttpClient.request",
                {"method": method, "url": url, "error": "timeout", "timeout": effective_timeout},
            )
            raise
        except Exception as e:
            logger.error("HttpClient.request", {"method": method, "url": url, "error": str(e)})
            raise

    async def _request_with_retry(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
        retry_kind: Optional[str] = None,
    ) -> Any:
        """Execute request with retry logic"""
        last_error: Optional[Exception] = None

        for attempt in range(self._retry_attempts):
            try:
                url = self._get_url() + path
                return await self._execute_with_retry429(url, method, body, request_timeout_millis, retry_kind)
            except Exception as error:
                last_error = error

                # Don't retry on client errors (4xx)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if 400 <= status < 500:
                        raise error

                # Wait before retry (except on last attempt)
                if attempt < self._retry_attempts - 1:
                    delay = (self._retry_delay_millis / 1000.0) * (2**attempt)
                    logger.warn(
                        "HttpClient.retry",
                        {
                            "method": method,
                            "path": path,
                            "attempt": attempt + 1,
                            "delay": delay,
                            "error": str(error),
                        },
                    )
                    await asyncio.sleep(delay)

        logger.error(
            "HttpClient.retry",
            {"method": method, "path": path, "error": "Max retries exceeded", "attempts": self._retry_attempts},
        )
        if last_error:
            raise last_error
        raise Exception("Max retries exceeded")

    async def _request_with_failover(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]],
        request_timeout_millis: Optional[int],
        affinity_key: Optional[str],
        retry_kind: Optional[str] = None,
    ) -> Any:
        """Execute request with failover logic"""
        if not self._load_balancer or not self._enable_failover:
            return await self._request_with_retry(method, path, body, request_timeout_millis, retry_kind)

        urls = self._load_balancer.get_all_urls()
        attempted_urls = set()
        last_error: Optional[Exception] = None

        logger.log(
            "HttpClient.failover",
            {"method": method, "path": path, "total_servers": len(urls), "affinity_key": affinity_key},
        )

        for _ in range(len(urls)):
            # Pass affinity key to load balancer for consistent routing
            url = self._load_balancer.get_next_url(affinity_key)

            if url in attempted_urls:
                continue

            attempted_urls.add(url)

            try:
                # 429s are retried in place (same backend, backoff-paced)
                # inside _execute_with_retry429 -- they are not a
                # backend-health signal, so they must not trigger failover
                # to a different server.
                result = await self._execute_with_retry429(
                    url + path, method, body, request_timeout_millis, retry_kind
                )

                # Mark backend as healthy on success
                self._load_balancer.mark_healthy(url)

                return result
            except Exception as error:
                last_error = error

                # Mark backend as unhealthy on failure (5xx or network errors)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if status >= 500:
                        self._load_balancer.mark_unhealthy(url)
                elif isinstance(error, (httpx.NetworkError, httpx.TimeoutException)):
                    self._load_balancer.mark_unhealthy(url)

                logger.warn(
                    "HttpClient.failover", {"url": url, "method": method, "path": path, "error": str(error)}
                )
                print(f"Request failed for {url}: {method} {path} - {error}")

                # Don't retry on client errors (4xx)
                if isinstance(error, httpx.HTTPStatusError):
                    status = error.response.status_code
                    if 400 <= status < 500:
                        raise error

                # Continue to next server for server errors or network issues

        logger.error(
            "HttpClient.failover",
            {"method": method, "path": path, "error": "All servers failed", "attempted": len(attempted_urls)},
        )
        if last_error:
            raise last_error
        raise Exception("All servers failed")

    def _get_url(self) -> str:
        """Get URL from load balancer or base URL"""
        if self._load_balancer:
            return self._load_balancer.get_next_url()
        return self._base_url or ""

    def get_load_balancer(self) -> Optional[LoadBalancer]:
        """Get load balancer instance"""
        return self._load_balancer

    async def close(self) -> None:
        """Close HTTP client and connection pool"""
        await self._client.aclose()

File: clients/client-laravel/src/Http/HttpClient.php
<?php

namespace Queen\Http;

use GuzzleHttp\Client;
use GuzzleHttp\Promise\PromiseInterface;
use GuzzleHttp\Promise\Utils as PromiseUtils;
use Psr\Http\Message\ResponseInterface;
use Queen\Exceptions\HttpException;

class HttpClient
{
    private ?string $baseUrl;
    private ?LoadBalancer $loadBalancer;
    private int $timeoutMillis;
    private int $retryAttempts;
    private int $retryDelayMillis;
    private bool $enableFailover;
    private ?string $bearerToken;
    private array $headers;
    private array $retry429;
    private Client $guzzle;

    public function __construct(array $options = [])
    {
        $this->baseUrl = $options['baseUrl'] ?? null;
        $this->loadBalancer = $options['loadBalancer'] ?? null;
        $this->timeoutMillis = $options['timeoutMillis'] ?? 30000;
        $this->retryAttempts = $options['retryAttempts'] ?? 3;
        $this->retryDelayMillis = $options['retryDelayMillis'] ?? 1000;
        $this->enableFailover = $options['enableFailover'] ?? true;
        $this->bearerToken = $options['bearerToken'] ?? null;
        $this->headers = $options['headers'] ?? [];
        // 429 (rate-limited) backoff policy, separate from the 5xx/network
        // retryAttempts above: ['maxAttempts' => int, 'baseMs' => int,
        // 'capMs' => int], all optional. See Retry429Policy.
        $this->retry429 = $options['retry429'] ?? [];

        // 'handler' overrides Guzzle's handler stack, the seam tests use to
        // drive the retry/failover paths without a live server.
        $handler = $options['handler'] ?? null;
        $this->guzzle = new Client($handler !== null ? ['handler' => $handler] : []);
    }

    // ===========================
    // Synchronous API
    // ===========================

    /**
     * $retryKind selects the 429 backoff budget: pass Retry429Policy::KIND_POP
     * for long-poll (wait=true) pop requests to get the unbounded policy,
     * null/anything else for the bounded one.
     */
    public function get(string $path, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        return $this->requestWithFailover('GET', $path, null, $requestTimeoutMillis, $affinityKey, $retryKind);
    }

    public function post(string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        return $this->requestWithFailover('POST', $path, $body, $requestTimeoutMillis, $affinityKey, $retryKind);
    }

    public function put(string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        return $this->requestWithFailover('PUT', $path, $body, $requestTimeoutMillis, $affinityKey, $retryKind);
    }

    public function delete(string $path, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        return $this->requestWithFailover('DELETE', $path, null, $requestTimeoutMillis, $affinityKey, $retryKind);
    }

    // ===========================
    // Async API (returns Guzzle promises)
    // ===========================
    //
    // Unlike the synchronous API these do NOT retry 429 in flight: the only
    // way to wait inside a promise chain here is to block, which would stall
    // every other request sharing the cURL multi-handle. The rejection is an
    // HttpException carrying errorCode/retryAfterSeconds, so callers pace
    // themselves between rounds (see ConsumerManager::concurrentWorkers) using
    // the policy from getRetry429Policy().

    public function getAsync(string $path, ?int $requestTimeoutMillis = null, ?string $affinityKey = null): PromiseInterface
    {
        return $this->executeRequestAsync($this->resolveUrl($affinityKey) . $path, 'GET', null, $requestTimeoutMillis);
    }

    public function postAsync(string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null): PromiseInterface
    {
        return $this->executeRequestAsync($this->resolveUrl($affinityKey) . $path, 'POST', $body, $requestTimeoutMillis);
    }

    public function putAsync(string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null): PromiseInterface
    {
        return $this->executeRequestAsync($this->resolveUrl($affinityKey) . $path, 'PUT', $body, $requestTimeoutMillis);
    }

    public function deleteAsync(string $path, ?int $requestTimeoutMillis = null, ?string $affinityKey = null): PromiseInterface
    {
        return $this->executeRequestAsync($this->resolveUrl($affinityKey) . $path, 'DELETE', null, $requestTimeoutMillis);
    }

    /**
     * Wait for multiple promises to resolve concurrently.
     *
     * @param PromiseInterface[] $promises
     * @return array Results indexed same as input
     */
    public static function awaitAll(array $promises): array
    {
        return PromiseUtils::unwrap($promises);
    }

    /**
     * Settle all promises (no exceptions on failure). Returns array of
     * ['state' => 'fulfilled'|'rejected', 'value' => ..., 'reason' => ...]
     *
     * @param PromiseInterface[] $promises
     * @return array
     */
    public static function settleAll(array $promises): array
    {
        return PromiseUtils::settle($promises)->wait();
    }

    // ===========================
    // Internals
    // ===========================

    public function getLoadBalancer(): ?LoadBalancer
    {
        return $this->loadBalancer;
    }

    /**
     * Effective 429 backoff policy for a request kind. Exposed so callers of
     * the async API — which has no in-flight 429 retry — pace their own poll
     * rounds with the same numbers the synchronous path uses.
     */
    public function getRetry429Policy(?string $retryKind = null): Retry429Policy
    {
        return Retry429Policy::forKind($this->retry429, $retryKind);
    }

    private function resolveUrl(?string $affinityKey = null): string
    {
        if ($this->loadBalancer !== null) {
            return $this->loadBalancer->getNextUrl($affinityKey);
        }
        if ($this->baseUrl === null) {
            throw new \LogicException('HttpClient has no baseUrl and no LoadBalancer configured');
        }
        return $this->baseUrl;
    }

    private function buildRequestOptions(string $method, ?array $body, ?int $requestTimeoutMillis): array
    {
        $effectiveTimeout = $requestTimeoutMillis ?? $this->timeoutMillis;

        $headers = ['Content-Type' => 'application/json'];
        if ($this->bearerToken !== null) {
            $headers['Authorization'] = "Bearer {$this->bearerToken}";
        }
        $headers = array_merge($headers, $this->headers);

        $options = [
            'headers' => $headers,
            'timeout' => $effectiveTimeout / 1000,
            'connect_timeout' => 5,
            'http_errors' => false,
        ];

        if ($body !== null) {
            $options['json'] = $body;
        }

        return $options;
    }

    private function parseResponse(ResponseInterface $response): mixed
    {
        $statusCode = $response->getStatusCode();

        if ($statusCode === 204) {
            return null;
        }

        $responseBody = (string) $response->getBody();

        if ($statusCode >= 400) {
            $error = "HTTP {$statusCode}";
            $errorCode = null;
            $reason = null;
            $detail = null;
            if ($responseBody) {
                $decoded = json_decode($responseBody, true);
                if (isset($decoded['error'])) {
                    $error = $decoded['error'];
                }
                // Proxy error contract: 429 {error, code: 'rate_limited' |
                // 'quota_exceeded'} with Retry-After (seconds); 403 {error,
                // code: 'cluster_suspended' | 'storage_quota_exceeded' |
                // 'feature_gated' | 'forbidden'}. See ErrorCode.
                if (isset($decoded['code']) && is_string($decoded['code'])) {
                    $errorCode = $decoded['code'];
                }
                // The kv/timers envelope carries two more fields: `reason`, a
                // finer stable identifier, and `detail`, which names the
                // offending operation index. Dropping them would leave the
                // caller with "kv_bad_request" and nothing to act on, on the
                // one surface whose ops arrive in batches.
                if (isset($decoded['reason']) && is_string($decoded['reason'])) {
                    $reason = $decoded['reason'];
                }
                if (isset($decoded['detail']) && is_string($decoded['detail'])) {
                    $detail = $decoded['detail'];
                }
            }

            $retryAfterSeconds = $statusCode === 429
                ? $this->parseRetryAfter($response->getHeaderLine('Retry-After'))
                : null;

            // The message stays the code so existing string-free branching is
            // unchanged, with the finer identifier and the human half appended
            // when the server sent them — a failing assertion that reads
            // "kv_bad_request" and nothing else costs an hour.
            $message = $error;
            if ($reason !== null && $reason !== $error) {
                $message .= ": {$reason}";
            }
            if ($detail !== null) {
                $message .= " ({$detail})";
            }

            throw new HttpException($message, $statusCode, 0, null, $errorCode, $retryAfterSeconds, $reason, $detail);
        }

        if (empty($responseBody)) {
            return null;
        }

        return json_decode($responseBody, true);
    }

    /**
     * Parse the Retry-After header (seconds, per the proxy contract) into a
     * float. Null when absent, non-numeric or negative.
     */
    private function parseRetryAfter(string $value): ?float
    {
        if ($value === '' || !is_numeric($value)) {
            return null;
        }

        $seconds = (float) $value;

        return $seconds >= 0 ? $seconds : null;
    }

    private function executeRequest(string $url, string $method, ?array $body = null, ?int $requestTimeoutMillis = null): mixed
    {
        $options = $this->buildRequestOptions($method, $body, $requestTimeoutMillis);
        $response = $this->guzzle->request($method, $url, $options);
        return $this->parseResponse($response);
    }

    /**
     * Run one logical request against a single URL, transparently retrying
     * HTTP 429 with backoff until the policy for $retryKind is exhausted (or
     * never, for the unbounded pop policy). Every other outcome — success,
     * network error, non-429 4xx, 5xx — passes straight through: 429 is the
     * only status this layer retries, and 5xx/network retry plus
     * cross-backend failover stay with the callers below.
     */
    private function executeRequestWithRetry429(string $url, string $method, ?array $body, ?int $requestTimeoutMillis, ?string $retryKind): mixed
    {
        $policy = Retry429Policy::forKind($this->retry429, $retryKind);
        $tries = 0;

        while (true) {
            $tries++;

            try {
                return $this->executeRequest($url, $method, $body, $requestTimeoutMillis);
            } catch (HttpException $error) {
                if ($error->statusCode !== 429 || $policy->isExhausted($tries)) {
                    throw $error;
                }

                usleep($policy->delayMillis($tries - 1, $error->retryAfterSeconds) * 1000);
            }
        }
    }

    private function executeRequestAsync(string $url, string $method, ?array $body = null, ?int $requestTimeoutMillis = null): PromiseInterface
    {
        $options = $this->buildRequestOptions($method, $body, $requestTimeoutMillis);

        return $this->guzzle->requestAsync($method, $url, $options)->then(
            fn(ResponseInterface $response) => $this->parseResponse($response)
        );
    }

    private function getStatusCode(\Throwable $error): int
    {
        return ($error instanceof HttpException) ? $error->statusCode : 0;
    }

    private function requestWithRetry(string $method, string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        $lastError = null;

        for ($attempt = 0; $attempt < $this->retryAttempts; $attempt++) {
            try {
                $url = $this->resolveUrl($affinityKey) . $path;
                return $this->executeRequestWithRetry429($url, $method, $body, $requestTimeoutMillis, $retryKind);
            } catch (\Throwable $error) {
                $lastError = $error;

                $statusCode = $this->getStatusCode($error);
                if ($statusCode >= 400 && $statusCode < 500) {
                    throw $error;
                }

                if ($attempt < $this->retryAttempts - 1) {
                    $delay = $this->retryDelayMillis * (2 ** $attempt);
                    usleep($delay * 1000);
                }
            }
        }

        throw $lastError;
    }

    private function requestWithFailover(string $method, string $path, ?array $body = null, ?int $requestTimeoutMillis = null, ?string $affinityKey = null, ?string $retryKind = null): mixed
    {
        if ($this->loadBalancer === null || !$this->enableFailover) {
            return $this->requestWithRetry($method, $path, $body, $requestTimeoutMillis, $affinityKey, $retryKind);
        }

        $urls = $this->loadBalancer->getAllUrls();
        $attemptedUrls = [];
        $lastError = null;

        for ($i = 0; $i < count($urls); $i++) {
            $url = $this->loadBalancer->getNextUrl($affinityKey);

            if (in_array($url, $attemptedUrls, true)) {
                continue;
            }

            $attemptedUrls[] = $url;

            try {
                // 429s are retried in place against this same backend inside
                // executeRequestWithRetry429: rate limiting is a tenant-quota
                // signal, not a backend-health one, so it must neither mark
                // the server unhealthy nor fail over to another (every backend
                // would answer the same, and spraying the fleet only makes the
                // limiter angrier). An exhausted 429 is a 4xx and therefore
                // leaves the loop below without a second server being tried.
                $result = $this->executeRequestWithRetry429($url . $path, $method, $body, $requestTimeoutMillis, $retryKind);
                $this->loadBalancer->markHealthy($url);
                return $result;
            } catch (\Throwable $error) {
                $lastError = $error;

                $statusCode = $this->getStatusCode($error);
                if ($statusCode === 0 || $statusCode >= 500) {
                    $this->loadBalancer->markUnhealthy($url);
                }

                if ($statusCode >= 400 && $statusCode < 500) {
                    throw $error;
                }
            }
        }

        throw $lastError ?? new \RuntimeException('All servers failed');
    }
}
